funsor/gaussian.py

```diff
@@ -498,11 +498,13 @@ def eager_reduce(self, op, reduced_vars):
         return None  # defer to default implementation
 
     def unscaled_sample(self, sampled_vars, sample_inputs):
-        # Sample only the real variables.
-        sampled_vars = frozenset(k for k, v in self.inputs.items()
-                                 if k in sampled_vars if v.dtype == 'real')
+        sampled_vars = sampled_vars.intersection(self.inputs)
         if not sampled_vars:
             return self
+        if any(self.inputs[k].dtype != 'real' for k in sampled_vars):
```

Comment:
After this PR we don't allow sampling discrete parts of a raw `Gaussian`; in practice this should never happen, any real distribution should pair a `Gaussian` with a normalizing `Tensor`.